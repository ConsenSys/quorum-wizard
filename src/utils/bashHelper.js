import { createDirectory, isTessera, networkFolder, includeCakeshop } from './networkCreator'
import {
  copyFile,
  createFolder,
  readFileToString,
  removeFolder,
  writeFile,
  writeJsonFile,
} from './fileUtils'
import { join, normalize } from 'path'
import { execute } from './execUtils'
import { buildCakeshopDir, generateCakeshopScript, waitForCakeshopCommand } from './cakeshopHelper'

export function buildBashScript(config) {
  const commands = createDirectory(config)

  const cakeshopStart = includeCakeshop(config) ? [generateCakeshopScript(), waitForCakeshopCommand()].join("") : ""

  const startScript = [
    commands.tesseraStart,
    waitForTesseraNodesCommand(config),
    commands.gethStart,
    cakeshopStart
  ]

  return {
    startScript: startScript.join('\n'),
    initCommands: commands.initStart,
    networkPath: commands.netPath
  }
}

export function buildBash(config) {

  const bashDetails = buildBashScript(config)
  const networkPath = bashDetails.networkPath

  if(includeCakeshop(config)) {
    buildCakeshopDir(config, join(networkPath, 'qdata'))
  }

  writeFile(join(networkPath, 'start.sh'), bashDetails.startScript, true)
  copyFile(join(process.cwd(), 'lib/stop.sh'), join(networkPath, 'stop.sh'))

  copyFile(join(process.cwd(), 'lib/runscript.sh'), join(networkPath, 'runscript.sh'))
  copyFile(join(process.cwd(), 'lib/public-contract.js'), join(networkPath, 'public-contract.js'))
  copyFile(join(process.cwd(), 'lib/private-contract.js'), join(networkPath, 'private-contract.js'))

  bashDetails.initCommands.forEach((command) => {
    // TODO figure out the safest way to run shell commands
    execute(command, (e, stdout, stderr) => {
      if (e instanceof Error) {
        console.error(e)
        throw e
      }
    })
  })
}

export function createGethStartCommand (config, node, passwordDestination, nodeNumber, tmIpcLocation) {
  const { verbosity, id, consensus } = config.network
  const { devP2pPort, rpcPort, wsPort, raftPort } = node.quorum

  const args = `--nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,${consensus},quorumPermission --emitcheckpoints --unlock 0 --password ${passwordDestination}`
  const consensusArgs = consensus === 'raft' ?
    `--raft --raftport ${raftPort}` :
    `--istanbul.blockperiod 5 --syncmode full --mine --minerthreads 1`

  return `PRIVATE_CONFIG=${tmIpcLocation} nohup geth --datadir qdata/dd${nodeNumber} ${args} ${consensusArgs} --permissioned --verbosity ${verbosity} --networkid ${id} --rpcport ${rpcPort} --port ${devP2pPort} 2>>qdata/logs/${nodeNumber}.log &`
}

export function createTesseraStartCommand (config, node, nodeNumber, tmDir, logDir) {
  // `rm -f ${tmDir}/tm.ipc`

  const tesseraJar = '$TESSERA_JAR' // require env variable to be set for now
  let DEBUG = ''
  if (config.network.remoteDebug) {
    DEBUG = '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=500$i -Xdebug'
  }

  const MEMORY = '-Xms128M -Xmx128M'
  const CMD = `java ${DEBUG} ${MEMORY} -jar ${tesseraJar} -configfile ${tmDir}/tessera-config-09-${nodeNumber}.json >> ${logDir}/tessera${nodeNumber}.log 2>&1 &`
  return CMD
}

export function waitForTesseraNodesCommand (config) {
  if(!isTessera(config)) {
    return ''
  }
  // TODO use config values for ip, port, data folder, etc.
  return `
echo "Waiting until all Tessera nodes are running..."
DOWN=true
k=10
while \${DOWN}; do
    sleep 1
    DOWN=false
    for i in \`seq 1 ${config.nodes.length}\`
    do
        if [ ! -S "qdata/c\${i}/tm.ipc" ]; then
            echo "Node \${i} is not yet listening on tm.ipc"
            DOWN=true
        fi

        set +e
        #NOTE: if using https, change the scheme
        #NOTE: if using the IP whitelist, change the host to an allowed host
        result=$(curl -s http://localhost:900\${i}/upcheck)
        set -e
        if [ ! "\${result}" == "I'm up!" ]; then
            echo "Node \${i} is not yet listening on http"
            DOWN=true
        fi
    done

    k=$((k - 1))
    if [ \${k} -le 0 ]; then
        echo "Tessera is taking a long time to start.  Look at the Tessera logs in qdata/logs/ for help diagnosing the problem."
    fi
    echo "Waiting until all Tessera nodes are running..."

    sleep 5
done

echo "All Tessera nodes started"
`
}
