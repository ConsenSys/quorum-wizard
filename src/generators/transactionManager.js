import { join } from 'path'
import { info } from '../utils/log'
import {
  cwd,
  readFileToString,
} from '../utils/fileUtils'
import { isTessera } from '../model/NetworkConfig'

// eslint-disable-next-line import/prefer-default-export
export function printTesseraKeys(config, log) {
  let keyString = !log ? 'echo "' : ''
  const qdata = join(cwd(), 'network', config.network.name, 'qdata')
  let pubKey = ''
  if (isTessera(config.network.transactionManager)) {
    keyString += '--------------------------------------------------------------------------------\n'
    config.nodes.forEach((node, i) => {
      const nodeNumber = i + 1
      pubKey = readFileToString(join(qdata, `c${nodeNumber}`, 'tm.pub'))
      keyString += `\nTessera Node ${nodeNumber} public key:\n${pubKey}\n`
    })
    keyString += '\n--------------------------------------------------------------------------------'
    keyString += !log ? '"\n' : '\n'
  }
  if (log) {
    info(keyString)
    return pubKey
  }
  return keyString
}
