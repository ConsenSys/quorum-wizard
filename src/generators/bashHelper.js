import {
  getFullNetworkPath,
} from './networkCreator'
import {
  copyFile,
  libRootDir,
  writeFile,
} from '../utils/fileUtils'
import { executeSync } from '../utils/execUtils'
import {
  buildCakeshopDir,
  generateCakeshopScript,
} from './cakeshopHelper'
import {
  isQuorum260Plus,
  pathToCakeshop,
  pathToQuorumBinary,
  pathToTesseraJar,
} from './binaryHelper'
import {
  isRaft,
  isTessera,
  isCakeshop,
} from '../model/NetworkConfig'
import { info } from '../utils/log'
import { formatTesseraKeysOutput } from './transactionManager'
import { joinPath } from '../utils/pathUtils'

export function buildBashScript(config) {
  const commands = createCommands(config)

  const startScript = [
    '#!/bin/bash',
    'echo "\nStarting Quorum network...\n"',
    setEnvironmentCommand(config),
    commands.tesseraStart,
    waitForTesseraNodesCommand(config),
    'echo "Starting Quorum nodes"',
    commands.gethStart,
    generateCakeshopScript(config),
    'echo "Successfully started Quorum network."',
    `echo "${formatTesseraKeysOutput(config)}"`,
  ]

  return {
    startScript: startScript.join('\n'),
    initCommands: commands.initStart,
  }
}

export function createCommands(config) {
  const networkPath = getFullNetworkPath(config)
  const initCommands = []
  const startCommands = []
  const tmStartCommands = []

  config.nodes.forEach((node, i) => {
    const nodeNumber = i + 1
    const quorumDir = joinPath('qdata', `dd${nodeNumber}`)
    const tmDir = joinPath('qdata', `c${nodeNumber}`)
    const genesisLocation = joinPath(quorumDir, 'genesis.json')
    const keyDir = joinPath(quorumDir, 'keystore')
    const passwordDestination = joinPath(keyDir, 'password.txt')
    const logs = joinPath('qdata', 'logs')
    const initCommand = `cd ${networkPath} && ${pathToQuorumBinary(config.network.quorumVersion)} --datadir ${quorumDir} init ${genesisLocation} 2>&1`
    initCommands.push(initCommand)

    const tmIpcLocation = isTessera(config.network.transactionManager)
      ? joinPath(tmDir, 'tm.ipc')
      : 'ignore'
    const startCommand = createGethStartCommand(
      config,
      node,
      passwordDestination,
      nodeNumber,
      tmIpcLocation,
    )
    startCommands.push(startCommand)

    if (isTessera(config.network.transactionManager)) {
      const tmStartCommand = createTesseraStartCommand(config, node, nodeNumber, tmDir, logs)
      tmStartCommands.push(tmStartCommand)
    }
  })

  const obj = {
    tesseraStart: tmStartCommands.join('\n'),
    gethStart: startCommands.join('\n'),
    initStart: initCommands,
  }
  return obj
}


export async function buildBash(config) {
  const bashDetails = buildBashScript(config)
  const networkPath = getFullNetworkPath(config)

  if (isCakeshop(config.network.cakeshop)) {
    buildCakeshopDir(config, joinPath(networkPath, 'qdata'))
  }

  info('Writing start script...')
  writeFile(joinPath(networkPath, 'start.sh'), bashDetails.startScript, true)
  copyFile(joinPath(libRootDir(), 'lib', 'stop.sh'), joinPath(networkPath, 'stop.sh'))

  info('Initializing quorum...')
  bashDetails.initCommands.forEach((command) => executeSync(command))
  info('Done')
}

export function createGethStartCommand(config, node, passwordDestination, nodeNumber, tmIpcPath) {
  const {
    verbosity, networkId, consensus, quorumVersion,
  } = config.network
  const {
    devP2pPort, rpcPort, raftPort, wsPort,
  } = node.quorum
  const quorum26Flags = isQuorum260Plus(quorumVersion) ? '--allow-insecure-unlock' : ''

  const args = `--nodiscover --rpc --rpccorsdomain=* --rpcvhosts=* --rpcaddr 0.0.0.0 --rpcapi admin,db,eth,debug,miner,net,shh,txpool,personal,web3,quorum,${consensus},quorumPermission --ws --wsaddr 0.0.0.0 --wsorigins=* --emitcheckpoints --unlock 0 --password ${passwordDestination} ${quorum26Flags}`
  const consensusArgs = isRaft(consensus)
    ? `--raft --raftport ${raftPort}`
    : '--istanbul.blockperiod 5 --syncmode full --mine --minerthreads 1'

  return `PRIVATE_CONFIG=${tmIpcPath} nohup $BIN_GETH --datadir qdata/dd${nodeNumber} ${args} ${consensusArgs} --permissioned --verbosity ${verbosity} --networkid ${networkId} --rpcport ${rpcPort} --wsport ${wsPort} --port ${devP2pPort} 2>>qdata/logs/${nodeNumber}.log &`
}

export function createTesseraStartCommand(config, node, nodeNumber, tmDir, logDir) {
  // `rm -f ${tmDir}/tm.ipc`

  let DEBUG = ''
  if (config.network.remoteDebug) {
    DEBUG = '-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=500$i -Xdebug'
  }

  const MEMORY = '-Xms128M -Xmx128M'
  const CMD = `java ${DEBUG} ${MEMORY} -jar $BIN_TESSERA -configfile ${tmDir}/tessera-config-09-${nodeNumber}.json >> ${logDir}/tessera${nodeNumber}.log 2>&1 &`
  return CMD
}

function checkTesseraUpcheck(nodes) {
  return nodes.map((node, i) => `
    result${i + 1}=$(curl -s http://${node.tm.ip}:${node.tm.p2pPort}/upcheck)
    if [ ! "\${result${i + 1}}" == "I'm up!" ]; then
        echo "Node ${i + 1} is not yet listening on http"
        DOWN=true
    fi`)
}

export function setEnvironmentCommand(config) {
  const lines = []
  lines.push(`BIN_GETH=${pathToQuorumBinary(config.network.quorumVersion)}`)
  if (isTessera(config.network.transactionManager)) {
    lines.push(`BIN_TESSERA=${pathToTesseraJar(config.network.transactionManager)}`)
  }
  if (isCakeshop(config.network.cakeshop)) {
    lines.push(`BIN_CAKESHOP=${pathToCakeshop(config.network.cakeshop)}`)
  }
  lines.push('')
  return lines.join('\n')
}

export function waitForTesseraNodesCommand(config) {
  if (!isTessera(config.network.transactionManager)) {
    return ''
  }
  return `
echo "Waiting until all Tessera nodes are running..."
DOWN=true
k=10
while \${DOWN}; do
    sleep 1
    DOWN=false
    for i in \`seq 1 ${config.nodes.length}\`
    do
        if [ ! -S "qdata/c\${i}/tm.ipc" ]; then
            echo "Node \${i} is not yet listening on tm.ipc"
            DOWN=true
        fi
    done
    set +e
    #NOTE: if using https, change the scheme
    #NOTE: if using the IP whitelist, change the host to an allowed host
    ${checkTesseraUpcheck(config.nodes).join('')}

    k=$((k - 1))
    if [ \${k} -le 0 ]; then
        echo "Tessera is taking a long time to start.  Look at the Tessera logs in qdata/logs/ for help diagnosing the problem."
    fi
    echo "Waiting until all Tessera nodes are running..."

    sleep 5
done

echo "All Tessera nodes started"
`
}
