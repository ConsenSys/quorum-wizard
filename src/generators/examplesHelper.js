import {
  copyFile,
  cwd, FILES,
  libRootDir,
  writeFile,
} from '../utils/fileUtils'
import { loadTesseraPublicKey } from './transactionManager'
import { isTessera } from '../model/NetworkConfig'
import { joinPath } from '../utils/pathUtils'
import {
  attachCommandBash,
  runscriptCommandBash,
  scriptHeader,
  validateNodeNumberInput,
} from './bashHelper'
import { isWin32 } from '../utils/execUtils'
import {
  kubernetesAttachCommand,
  kubernetesRunscriptCommand,
} from './kubernetesHelper'
import {
  dockerAttachCommand,
  dockerRunscriptCommand,
} from './dockerHelper'

export function generateAndCopyExampleScripts(config) {
  const networkPath = joinPath(cwd(), 'network', config.network.name)
  writeFile(joinPath(networkPath, FILES.runscript), generateRunScript(config), true)
  writeFile(joinPath(networkPath, FILES.attach), generateAttachScript(config), true)
  copyFile(
    joinPath(libRootDir(), 'lib', FILES.publicContract),
    joinPath(networkPath, FILES.publicContract),
  )
  if (isTessera(config.network.transactionManager)) {
    const nodeTwoPublicKey = loadTesseraPublicKey(config, 2)
    writeFile(joinPath(networkPath, FILES.privateContract),
      generatePrivateContractExample(nodeTwoPublicKey))
  }
}

function generatePrivateContractExample(privateFor) {
  return `
a = eth.accounts[0]
web3.eth.defaultAccount = a;

// abi and bytecode generated from simplestorage.sol:
// > solcjs --bin --abi simplestorage.sol
var abi = [{"constant":true,"inputs":[],"name":"storedData","outputs":[{"name":"","type":"uint256"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"x","type":"uint256"}],"name":"set","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"get","outputs":[{"name":"retVal","type":"uint256"}],"payable":false,"type":"function"},{"inputs":[{"name":"initVal","type":"uint256"}],"payable":false,"type":"constructor"}];

var bytecode = "0x6060604052341561000f57600080fd5b604051602080610149833981016040528080519060200190919050505b806000819055505b505b610104806100456000396000f30060606040526000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680632a1afcd914605157806360fe47b11460775780636d4ce63c146097575b600080fd5b3415605b57600080fd5b606160bd565b6040518082815260200191505060405180910390f35b3415608157600080fd5b6095600480803590602001909190505060c3565b005b341560a157600080fd5b60a760ce565b6040518082815260200191505060405180910390f35b60005481565b806000819055505b50565b6000805490505b905600a165627a7a72305820d5851baab720bba574474de3d09dbeaabc674a15f4dd93b974908476542c23f00029";

var simpleContract = web3.eth.contract(abi);
var simple = simpleContract.new(42, {from:web3.eth.accounts[0], data: bytecode, gas: 0x47b760, privateFor: ["${privateFor}"]}, function(e, contract) {
  if (e) {
    console.log("err creating contract", e);
  } else {
    if (!contract.address) {
      console.log("Contract transaction send: TransactionHash: " + contract.transactionHash + " waiting to be mined...");
    } else {
      console.log("Contract mined! Address: " + contract.address);
      console.log(contract);
    }
  }
});`
}

function getAttachCommand(config) {
  switch (config.network.deployment) {
    case 'bash':
      return attachCommandBash(config)
    case 'docker-compose':
      return dockerAttachCommand()
    case 'kubernetes':
      return kubernetesAttachCommand()
    default:
      return ''
  }
}

export function generateAttachScript(config) {
  const command = getAttachCommand(config)
  return `${scriptHeader()}
${validateNodeNumberInput(config)}
${command}`
}

function generateRunscriptCommand(config) {
  switch (config.network.deployment) {
    case 'bash':
      return runscriptCommandBash(config)
    case 'docker-compose':
      return dockerRunscriptCommand()
    case 'kubernetes':
      return kubernetesRunscriptCommand()
    default:
      return ''
  }
}

export function generateRunScript(config) {
  const command = generateRunscriptCommand(config)
  return `${scriptHeader()}
${filenameCheck()}
${command}
`
}

function filenameCheck() {
  return isWin32() ? filenameCheckWindows() : filenameCheckBash()
}

function filenameCheckWindows() {
  return `if NOT "%1"=="${FILES.privateContract}" if NOT "%1"=="${FILES.publicContract}" (
  echo Please provide a valid script file to execute (i.e. ${FILES.run}${FILES.runscript} ${FILES.privateContract}) && EXIT /B 1
)`
}

function filenameCheckBash() {
  return `if [ -z $1 ] || [ ! -f $1 ]; then
  echo "Please provide a valid script file to execute (i.e. ${FILES.run}${FILES.runscript} ${FILES.privateContract})" >&2
  exit 1
fi`
}

