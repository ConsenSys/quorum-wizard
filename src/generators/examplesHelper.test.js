import {
  createNetPath,
  createLibPath,
  TEST_CWD,
  TEST_LIB_ROOT_DIR,
} from '../utils/testHelper'
import {
  copyFile,
  getOutputPath,
  libRootDir,
  writeFile,
} from '../utils/fileUtils'
import { generateAndCopyExampleScripts, generateAttachScript, generateRunScript } from './examplesHelper'
import { loadTesseraPublicKey } from './transactionManager'
import { setEnvironmentCommand } from './bashHelper'

jest.mock('./bashHelper')
jest.mock('../utils/fileUtils')
jest.mock('./transactionManager')
getOutputPath.mockReturnValue(TEST_CWD)
libRootDir.mockReturnValue(TEST_LIB_ROOT_DIR)
loadTesseraPublicKey.mockReturnValue('publickey')
setEnvironmentCommand.mockReturnValue('BIN_GETH=path/to/binary')

const CONFIG = {
  network: {
    name: 'test',
    networkPath: TEST_CWD,
  },
  nodes: [{
    quorum: {
      ip: '1.2.3.4',
      rpcPort: '1234',
    },
  }],
}

describe('generates and copies over example scripts', () => {
  it('generates private-contract with node 2 and copies over the 3 example scripts', () => {
    generateAndCopyExampleScripts(CONFIG)
    expect(writeFile).toBeCalledWith(
      createNetPath(CONFIG, 'runscript.sh'),
      expect.anything(),
      true,
    )
    expect(writeFile).toBeCalledWith(
      createNetPath(CONFIG, 'attach.sh'),
      expect.anything(),
      true,
    )
    expect(copyFile).toBeCalledWith(
      createLibPath('lib', 'public_contract.js'),
      createNetPath(CONFIG, 'public_contract.js'),
    )
    expect(writeFile).toBeCalledWith(
      createNetPath(CONFIG, 'private_contract.js'),
      expect.anything(),
    )
    expect(loadTesseraPublicKey).toBeCalledWith(CONFIG, 2)
  })
  it('generates public-contract and copies over the 2 example scripts', () => {
    const config = {
      ...CONFIG,
      network: {
        ...CONFIG.network,
        transactionManager: 'none',
      },
    }
    generateAndCopyExampleScripts(config)
    expect(writeFile).toBeCalledWith(
      createNetPath(config, 'runscript.sh'),
      expect.anything(),
      true,
    )
    expect(copyFile).toBeCalledWith(
      createLibPath('lib', 'public_contract.js'),
      createNetPath(config, 'public_contract.js'),
    )
  })
  it('generates the runscript and attach shell scripts for bash', () => {
    const config = {
      ...CONFIG,
      network: {
        ...CONFIG.network,
        deployment: 'bash',
      },
    }
    expect(generateRunScript(config)).toMatchSnapshot()
    expect(generateAttachScript(config)).toMatchSnapshot()
  })
  it('generates the runscript and attach shell scripts for docker', () => {
    const config = {
      ...CONFIG,
      network: {
        ...CONFIG.network,
        deployment: 'docker-compose',
      },
    }
    expect(generateRunScript(config)).toMatchSnapshot()
    expect(generateAttachScript(config)).toMatchSnapshot()
  })
  it('generates the runscript and attach shell scripts for kubernetes', () => {
    const config = {
      ...CONFIG,
      network: {
        ...CONFIG.network,
        deployment: 'kubernetes',
      },
    }
    expect(generateRunScript(config)).toMatchSnapshot()
    expect(generateAttachScript(config)).toMatchSnapshot()
  })
})
