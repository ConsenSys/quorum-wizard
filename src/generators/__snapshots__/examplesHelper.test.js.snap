// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates and copies over example scripts generates the runscript and attach shell scripts for bash 1`] = `
"#!/bin/bash
if [ -z $1 ] || [ ! -f $1 ]; then
  echo \\"Please provide a valid script file to execute (i.e. ./runscript.sh private_contract.js)\\" >&2
  exit 1
fi
BIN_GETH=path/to/binary
BIN_TESSERA=path/to/binary
BIN_CAKESHOP=path/to/binary

$BIN_GETH --exec \\"loadScript(\\\\\\"$1\\\\\\")\\" attach qdata/dd1/geth.ipc
"
`;

exports[`generates and copies over example scripts generates the runscript and attach shell scripts for bash 2`] = `
"#!/bin/bash
NUMBER_OF_NODES=1
NODE_NUMBER=$1
case \\"$NODE_NUMBER\\" in (\\"\\" | *[!0-9]*)
  echo 'Please provide the number of the node to attach to (i.e. ./attach.sh 2)' >&2
  exit 1
esac

if [ \\"$NODE_NUMBER\\" -lt 1 ] || [ \\"$NODE_NUMBER\\" -gt $NUMBER_OF_NODES ]; then
  echo \\"$NODE_NUMBER is not a valid node number. Must be between 1 and $NUMBER_OF_NODES.\\" >&2
  exit 1
fi
BIN_GETH=path/to/binary
BIN_TESSERA=path/to/binary
BIN_CAKESHOP=path/to/binary

$BIN_GETH attach qdata/dd$1/geth.ipc"
`;

exports[`generates and copies over example scripts generates the runscript and attach shell scripts for docker 1`] = `
"#!/bin/bash
if [ -z $1 ] || [ ! -f $1 ]; then
  echo \\"Please provide a valid script file to execute (i.e. ./runscript.sh private_contract.js)\\" >&2
  exit 1
fi
docker cp $1 \\"$(docker-compose ps -q node1)\\":/$1
docker-compose exec node1 /bin/sh -c \\"geth --exec 'loadScript(\\\\\\"$1\\\\\\")' attach qdata/dd/geth.ipc\\"

"
`;

exports[`generates and copies over example scripts generates the runscript and attach shell scripts for docker 2`] = `
"#!/bin/bash
NUMBER_OF_NODES=1
NODE_NUMBER=$1
case \\"$NODE_NUMBER\\" in (\\"\\" | *[!0-9]*)
  echo 'Please provide the number of the node to attach to (i.e. ./attach.sh 2)' >&2
  exit 1
esac

if [ \\"$NODE_NUMBER\\" -lt 1 ] || [ \\"$NODE_NUMBER\\" -gt $NUMBER_OF_NODES ]; then
  echo \\"$NODE_NUMBER is not a valid node number. Must be between 1 and $NUMBER_OF_NODES.\\" >&2
  exit 1
fi
docker-compose exec node$1 /bin/sh -c \\"geth attach qdata/dd/geth.ipc\\""
`;

exports[`generates and copies over example scripts generates the runscript and attach shell scripts for kubernetes 1`] = `
"#!/bin/bash
if [ -z $1 ] || [ ! -f $1 ]; then
  echo \\"Please provide a valid script file to execute (i.e. ./runscript.sh private_contract.js)\\" >&2
  exit 1
fi

NODE_NUMBER=1
POD=$(kubectl get pod --field-selector=status.phase=Running -o name | grep quorum-node$NODE_NUMBER)
kubectl $NAMESPACE exec -it $POD -c quorum -- /etc/quorum/qdata/contracts/runscript.sh /etc/quorum/qdata/contracts/$1
"
`;

exports[`generates and copies over example scripts generates the runscript and attach shell scripts for kubernetes 2`] = `
"#!/bin/bash
NUMBER_OF_NODES=1
NODE_NUMBER=$1
case \\"$NODE_NUMBER\\" in (\\"\\" | *[!0-9]*)
  echo 'Please provide the number of the node to attach to (i.e. ./attach.sh 2)' >&2
  exit 1
esac

if [ \\"$NODE_NUMBER\\" -lt 1 ] || [ \\"$NODE_NUMBER\\" -gt $NUMBER_OF_NODES ]; then
  echo \\"$NODE_NUMBER is not a valid node number. Must be between 1 and $NUMBER_OF_NODES.\\" >&2
  exit 1
fi
POD=$(kubectl get pod --field-selector=status.phase=Running -o name | grep quorum-node$NODE_NUMBER)
kubectl $NAMESPACE exec -it $POD -c quorum -- /geth-helpers/geth-attach.sh"
`;
